src/main/scala/io/picnicml/doddlemodel/cluster/DBSCAN.scala

```diff
@@ -0,0 +1,85 @@
+package io.picnicml.doddlemodel.cluster
+
+import breeze.linalg.functions.euclideanDistance
+import cats.syntax.option._
+import io.picnicml.doddlemodel.data.Features
+import io.picnicml.doddlemodel.typeclasses.Clusterer
+
+/** An immutable DBSCAN model.
+  *
+  * @param eps: the maximum distance between points in a group
+  * @param minSamples: the minimum number of point in a core group
+  *
+  * Examples:
+  * val model = DBSCAN()
+  * val model = DBSCAN(eps = 1.5)
+  * val model = DBSCAN(minSamples = 3)
+  * val model = DBSCAN(eps = 2.0, minSamples = 3)
+  */
+case class DBSCAN private(eps: Double, minSamples: Int, private val label: Option[Array[Int]])
+
+object DBSCAN {
+
+  def apply(eps: Double = 1.0, minSamples: Int = 1): DBSCAN = {
+    require(eps > 0.0, "Maximum distance needs to be larger than 0")
+    require(minSamples > 0, "Minimum number of samples needs to be larger than 0")
+    DBSCAN(eps, minSamples, none)
+  }
+
+  implicit lazy val ev: Clusterer[DBSCAN] = new Clusterer[DBSCAN] {
+
+    override def isFitted(model: DBSCAN): Boolean = model.label.isDefined
+
+    override protected def labelSafe(model: DBSCAN): Array[Int] = model.label.get
+
+    override protected def copy(model: DBSCAN): DBSCAN =
+      model.copy()
+
+    override protected def copy(model: DBSCAN, label: Array[Int]): DBSCAN =
+      model.copy(label = label.some)
+
+    override protected def fitSafe(model: DBSCAN, x: Features): DBSCAN = {
+      val xSize = x.rows
+      val label = Array.fill[Int](xSize)(Int.MaxValue)
+      var groupStartId = 0
+      var groupId = 0
+      while (groupStartId < xSize) {
+        var groupQueue = Set(groupStartId)
+        var groupCount = 1
+        label(groupStartId) = groupId
+        while (groupQueue.size > 0) {
+          val tmpGroupQueue = groupQueue
+          groupQueue = Set[Int]()
+          for (pointId <- tmpGroupQueue; i <- 0 until xSize
+            if label(i) > groupId &&
+              euclideanDistance(x(i, ::).t, x(pointId, ::).t) <= model.eps) {
+            label(i) = groupId
+            groupQueue += i
+            groupCount += 1
+          }
+        }
+        if (groupCount < model.minSamples) {
+          for (i <- 0 until xSize if label(i) == groupId) {
+            label(i) = -1
+          }
+          groupId -= 1
+        }
+        groupStartId = findUnidentifiedPoint(label)
+        groupId += 1
+      }
+      copy(model, label)
+    }
+
+    override protected def fitPredictSafe(model: DBSCAN, x: Features): Array[Int] = {
+      labelSafe(fitSafe(model, x))
+    }
+
+    private def findUnidentifiedPoint(label: Array[Int]): Int = {
```

Comment:
Does this method bring a significant improvement in efficiency (or serves a specific purpose)? 
If it does not, I think this could be removed and the outer-most while loop could be replaced by an iteration through all points, thereby getting rid of some mutable constructs.

So the outer loop in `fitSafe` would look something like this:
```scala
// .view makes sure things get generated on the fly, instead of creating the whole sequence in advance
(0 until xSize).view.foreach {
    case idxExample if label(idxExample) == Int.MaxValue {
    // the code inside would need some modifications as well to work with this new approach
    // ...
    }
}
```

The reasons why I think this could work is that you go through each point once both with your way and with this new approach.

_Perhaps an even better approach_ would be using `foldLeft` instead of `foreach` and storing the `groupId` in the accumulator, thereby getting rid of another `var`.

Reply:
Purely based on the complexity of this method could introduce to DBSCAN, I think it has 2 possible benefits.
1. a major time saves: since the points that are able to actively connect to others are limited, it could be a lot faster if a good minSample is selected.
2. the possibility to decouple the problem (to scale): if there is no minSample, all points are treated equally. There is no hierarchy in the data points so all points might link to all other points. With minSample being introduced, there are 2 different levels of data points, and we are able to do a CORE-LEVEL vs NON-CORE-LEVEL connection parallelly like how we had been doing for k-Means.

Reply:
Okay, I'd need to think a bit more about these cases, but quickly thinking about this, it does seem possible.

In any way, you can write this in a more functional manner using `indexWhere`, using something like
```scala
(0 until label.length).view.indexWhere(i => label(i) == Int.MaxValue)
```
but note that `-1` is returned if no element satisfies the condition.

One more question about this, do you need to start the search from 0 every time? In the main loop, you are going through indices from 0 up to number of points and checking if they satisfy the core point condition. So they always get assigned to either **noise** or **core point**. This assignment might get changed at a later stage, but they will never again be unassigned.
So if the above is correct and I didn't miss something, I think you could start searching from `groupStartId`.

Reply:
which means instead of using while, we can simply use a for loop and remove the groupStartId variable.
The code is thereby modified.